qualityattenuation/qualityattenuation.go

```diff
@@ -0,0 +1,133 @@
+// Implements a data structure for quality attenuation.
+
+package qualityattenuation
+
+import (
+	"math"
+
+	"github.com/influxdata/tdigest"
+)
+
+type SimpleQualityAttenuation struct {
+	EmpiricalDistribution     *tdigest.TDigest
+	Offset                    float64
+	Offset_sum                float64
+	Offset_sum_of_squares     float64
+	Number_of_samples         int64
+	Number_of_losses          int64
+	Latency_eq_loss_threshold float64
+	Minimum_latency           float64
+	Maximum_latency           float64
+}
+
+func NewSimpleQualityAttenuation() *SimpleQualityAttenuation {
+	return &SimpleQualityAttenuation{
+		EmpiricalDistribution:     tdigest.NewWithCompression(50),
+		Offset:                    0.1,
+		Offset_sum:                0.0,
+		Offset_sum_of_squares:     0.0,
+		Number_of_samples:         0,
+		Number_of_losses:          0,
+		Latency_eq_loss_threshold: 15.0, // Count latency greater than this value as a loss.
+		Minimum_latency:           0.0,
+		Maximum_latency:           0.0,
+	}
+}
+
+func (qa *SimpleQualityAttenuation) AddSample(sample float64) {
+	if sample <= 0.0 {
+		// Ignore zero or negative samples because they cannot be valid.
+		// TODO: This should raise a warning and/or trigger error handling.
+		return
+	}
+	qa.Number_of_samples++
```

Comment:
In go, when you implement the member function you can have access to the private fields of the struct. 

TL;DR: You are using *excellent* go coding without even knowing it!!